File: codeforces/telepanting.py
#MLE on tc6

from bisect import bisect_right

MOD = 998244353

n = int(input())
p, xs = {}, []
for i in range(n):
    x, y, a = map(int, input().split())
    p[x] = (y, a)
    xs.append(x)

xs.sort()
dp = [0]*(n+1)
cur, ans = 0, xs[-1]+1
while cur < xs[-1]:
    pidx = bisect_right(xs, cur)
    next = xs[pidx]
    ncur = p[next][0]
    npidx = bisect_right(xs, ncur-1)
    nnext = xs[npidx]
    step = next-ncur+dp[pidx]-dp[npidx]
    if p[next][1]: ans += step
    dp[pidx+1] = step+dp[pidx]
    cur = next; ans %= MOD

print(ans)

File: codeforces/telepanting.cpp
// AC

#include <algorithm>
#include <bits/stdc++.h>
#include <utility>
#include <vector>
using namespace std;

#define int long long

const int MOD = 998244353 ;

signed main() {
    int n; cin >> n;
    unordered_map<int, pair<int, int>> p;
    vector<int> xs;
    for(int i=0; i<n; i++) {
        int x, y, a; cin >> x >> y >> a;
        p[x] = make_pair(y, a);
        xs.push_back(x);
    }

    sort(xs.begin(), xs.end());
    vector<int> dp(n+1);
    int cur = 0, ans = xs[n-1]+1;
    while(cur < xs[n-1]) {
        int pidx = distance(xs.begin(), upper_bound(xs.begin(), xs.end(), cur));
        int next = xs[pidx];
        int ncur = p[next].first;
        int npidx = distance(xs.begin(), lower_bound(xs.begin(), xs.end(), ncur));
        int nnext = xs[npidx];
        int step = (next-ncur+dp[pidx]-dp[npidx])%MOD;
        if(p[next].second) ans += step;
        dp[pidx+1] = (step+dp[pidx])%MOD;
        cur = next;
        ans %= MOD;
    }

    if(ans < 0) ans += MOD;

    cout << ans << endl;
    return 0;
}
